### DISCUSSION TITLE
How to differentiate when to use __css and sx props when building theme

### DISCUSSION THREAD

**Discussion by dundja** [chakra-ui/chakra-ui#5182] [Q&A]
Hey all,

We have an issue with stying our Text component with `__css` prop using useMultiStyleConfig hook. Basically we are creating our custom theme which we want to be extendable. Therefore we have created multi style part component that has `Box` component as wrapper and inside we want to use just two `Text` components.

Our question is when should we use `__css` and when should we use `sx` prop for stying those components since we saw that we can't style `Text` component with `__css` prop ?

```javascript
const showDataTheme = {
  parts: ['wrapper', 'label', 'data'],
  baseStyle: {
    wrapper: {
      justifyContent: 'flex-start',
      alignItems: 'flex-start',
      width: '100%',
    },
    label: {
      fontWeight: 700,
    },
    data: {
      color: 'blue',
    },
  },
};
```

So as you can see from this snippet, `label` styles are not applied to Text component with `__css` prop but work with `sx` prop. But if we change component from `Text` to `Box` styles are correctly applied with `__css` prop.

Many thanks.

**Comment by primos63** [comment]
`__css` has the lowest precedence and `sx` has the highest. `__css` should be used when you're creating a new component. `sx` should be used on your tag as a prop which will override/add to styles set in the component.
> So as you can see from this snippet, label styles are not applied to Text component with __css prop but work with sx prop. But if we change component from Text to Box styles are correctly applied with __css prop.

Actually I don't see as you only provided the styling for the component. If you could include the source for your component, then maybe I could see what's going on.

**Reply by dundja** [reply]
@primos63 thanks for your answer.

It is not about who precedence, it is about that other components, except `Box`, are not applying styles when you pass them via `__css` prop.

So we have our `label` component
```javascript
const Label = ({
  children,
  variant,
  ...props
}) => {
  const styles = useMultiStyleConfig('ShowData', { variant });

  return (
    <FormLabel __css={styles.label} {...props}>{children}</FormLabel>
  );
};
```
and we have our `data` component
```javascript
const Data = ({
  variant,
  children,
  ...props
}) => {
  const styles = useMultiStyleConfig('ShowData', { variant });

  return (
    <Box __css={styles.data} {...props}>{children}</Box>
  );
};
```

So now when we try to style them inside theme, styles wont apply to our `label` component but works perfectly fine with `data` component
```javascript
const theme = {
 ...restOfTheTheme
  components: {
              ShowData: {
                baseStyle: {
                  label: {
                    fontWeight: 700,
                    fontSize: 'lg',
                    color: 'blue.800',
                  },
                  data: {
                    marginTop: 2,
                    color: 'blue.500',
                    fontSize: 'md',
                    whiteSpace: 'nowrap',
                  },
                },
              },
            },
}
```

**Comment by primos63** [comment]
It is about precedence and understanding how a component works. Let's take the `Label` component.
```jsx
const Label = ({
  children,
  variant,
  ...props
}) => {
  const styles = useMultiStyleConfig('ShowData', { variant });

  return (
    <FormLabel __css={styles.label} {...props}>{children}</FormLabel>
  );
};
```

Here `styles` will be
![image](https://user-images.githubusercontent.com/2768849/144422097-156ad17d-0ba0-4b0d-943e-604bbdbf6cb2.png)
So if we use `<Label>Full name</Label>` in an app, then `__css` will be `fontWeight: 700` and `props` will be `{}`.
Now we render FormLabel, `<FormLabel __css={styles.label} {...props}>{children}</FormLabel>`. The code for FormLabel is
```jsx
export const FormLabel = forwardRef<FormLabelProps, "label">(
  (passedProps, ref) => {
    const styles = useStyleConfig("FormLabel", passedProps)
    const props = omitThemingProps(passedProps)

    const {
      className,
      children,
      requiredIndicator = <RequiredIndicator />,
      ...rest  
    } = props

    const field = useFormControlContext()
    const ownProps = field?.getLabelProps(rest, ref) ?? { ref, ...rest }

    return (
      <chakra.label
        {...ownProps}   
        className={cx("chakra-form__label", props.className)}
        __css={{
          display: "block",
          textAlign: "start",
          ...styles,
        }}
      >
        {children}
        {field?.isRequired ? requiredIndicator : null}
      </chakra.label>
    )
  },
)
```
Here's what the FormLabel's `styles`, `props`, `rest` and `ownProps` contain
![image](https://user-images.githubusercontent.com/2768849/144424171-ca0ff054-e0aa-4bdc-88d7-2fde26b10ec8.png)
Now the important part, we render `chakra.div`. This is what that will look like.
```jsx
<chakra.div
    ref=null
    __css={{ fontWeight: 700 }}
    className="chakra-form__label" + any other names in props.className
    __css={{
      display: "block",
      textAlign: "start",
      // ...styles from FormLabel only listing the attribute of interest
      fontWeight: "medium"
      ...
    }}>
```
So what happens here is that `Label`'s `__css` is replaced by `FormLabel`'s `__css` and the `fontWeight` appears not to change from `md` to `700`. Now let's change the rendering in `Label` to
```jsx
 return (
   <FormLabel {...styles.label} {...props}>{children}</FormLabel>
 );
```
The `chakra.div` is now rendered with
```jsx
<chakra.div
    ref=null
    fontWeight: 700
    className="chakra-form__label" + any other names in props.className
    __css={{
      display: "block",
      textAlign: "start",
      // ...styles from FormLabel only listing the attribute of interest
      fontWeight: "medium"
      ...
    }}>
```
As `__css` has the lowest precedence, it's processed first. The `props` are then processed and merged with those provided by `__css`. Now the `fontWeight` from `__css` is overridden with that from `Label`'s style. We can see that in the devtools by inspecting the `Label` element and seeing that `font-weight` is now `700` rather than `md`.

Why doesn't this happen with `Box`? Well `Box` has no styles it's simply a `div` so using the `__css` prop works as there is nothing in the theme which will override the properties in `__css`.
